src/intel_hex.nim

```diff
@@ -0,0 +1,263 @@
+            (hag.addrRow.bytes[1].uint32 shl 16))
+
+proc toByteList*(hag: HexAddrGroup): seq[AddressedByte] =
+  result = @[]
+  let baseOffset = hag.getBaseOffset()
+  for row in hag.dataRows:
+    if row.rowType != rtData:
+      continue
+    var rowOffset = row.offset
+    for b in row.bytes:
+      result.add(AddressedByte(address: baseOffset or rowOffset, data: b))
+      inc(rowOffset)
+
+proc toU32List*(hf: HexFile): seq[AddressedU32] =
+  result = @[]
+  var byteList = newSeq[AddressedByte]()
+
+  for ag in hf.addrGroups:
+    byteList &= ag.toByteList()
+
+  for chunk in byteList.chunksOf(4):
+    let wordOffset = chunk[0].address
+    var word: uint32 = 0
+    for i in 0..chunk.high:
+      word = word or (chunk[i].data shl (8 * (3 - i)))
+
+    result.add(AddressedU32(address: wordOffset, data: word))
+
+proc newHexAddrGroup*(): HexAddrGroup =
+  new(result)
+  result.dataRows = @[]
+
+proc newHexAddrGroup*(addrRow: HexRow): HexAddrGroup =
+  result = newHexAddrGroup()
+  result.addrRow = addrRow
+
+proc `+=`*(hag: HexAddrGroup, hr: HexRow) =
+  hag.dataRows.add(hr)
+
+proc `+=`*(hf: HexFile, hag: HexAddrGroup) =
+  hf.addrGroups.add(hag)
+
+proc `+=`*(hf: HexFile, hr: HexRow) =
+  hf.addrGroups[^1] += hr
+
+proc newHexFile(): HexFile =
+  new(result)
+  result.addrGroups = @[]
+
+proc fromHexFile*(path: string): HexFile =
+  result = newHexFile()
+
+  for line in lines(path):
+    let row: HexRow = fromStr(line)
+
+    if row.rowType == rtExtLinAddr:
+      let ag = newHexAddrGroup(row)
+      result.addrGroups.add(ag)
+    else:
+      # For things like AVR which only have a 16 bit address
+      # space and therefore don't need to place their code
+      # at a deep offset, we need to create a default
+      # address group starting at 0
+      if len(result.addrGroups) == 0:
+        let defaultRow = fromStr(DefaultAddrGroup)
+        let ag = newHexAddrGroup(defaultRow)
+        result.addrGroups.add(ag)
+      result.addrGroups[^1] += row
+
+proc addAddrGroupFromOffset(hf: HexFile, offset: uint16) =
+  let ar = fromBytes(
+    [(offset shr 8).byte, (offset and 0xff).byte],
+    0,
+    rtExtLinAddr
+  )
+  hf += newHexAddrGroup(ar)
+
+
+proc fromBinaryFile*(path: string, baseOffset: uint32): HexFile =
+  result = newHexFile()
+
+  let binData = readFile(path).map(proc (x: char):byte = x.byte)
+
+  var offset = (baseOffset and 0xffff).uint32
+  var baseOffset = (baseOffset shr 16).uint16
+
+  result.addAddrGroupFromOffset(baseOffset)
+
+  baseOffset += 1
+  var startIdx: uint32 = 0
+
+  while startIdx < len(binData).uint32:
+    let remainingSpace = AddrSpaceSize - offset
+    if remainingSpace == 0:
+      result.addAddrGroupFromOffset(baseOffset)
+      offset = 0
+      inc(baseOffset)
+      continue
+    let endIdx = min(startIdx + remainingSpace, len(binData).uint32)
+    let workingBlock = binData[startIdx..<endIDX]
+    for chunk in workingBlock.chunksOf(RowChunkSize):
```

Comment:
A note on the assignment of `workingBlock` here.  My first pass at this used the `binData[startIdx..<endIDX]` slice in place of the assigned name and the library was SO SLOW.  After some profiling, I realized that this line was at fault, and the presumed (I've not been able to fully prove it out yet, but I'm going to try) reason is that the slice operation was performed on every go-around of the loop, causing an immense amount of heap thrash and unnecessary bounds checking.  I believe that this is a consequence of how nim expanded the generic openarray.